tests/test_pipeline/test_pipeline.py

```diff
@@ -356,3 +362,130 @@ def test_forecast_pipeline_with_nan_at_the_end(df_with_nans_in_tails):
+
+
+@pytest.mark.parametrize(
+    "mask", (FoldMask(None, "2020-01-02", ["2020-01-03"]), FoldMask(None, "2020-01-05", ["2020-01-06"]))
+)
+def test_generate_folds_datasets_without_first_date(simple_ts: TSDataset, mask):
+    """Check _generate_folds_datasets for correct work without first date."""
+    pipeline = Pipeline(model=NaiveModel(lag=7))
+    mask = pipeline._prepare_fold_masks(ts=simple_ts, masks=[mask], mode="constant")[0]
+    train, test = list(pipeline._generate_folds_datasets(simple_ts, [mask], 4))[0]
+    assert train.index.min() == np.datetime64(simple_ts.index.min())
+    assert train.index.max() == np.datetime64(mask.last_train_timestamp)
+    assert test.index.min() == np.datetime64(mask.last_train_timestamp) + np.timedelta64(1, "D")
+    assert test.index.max() == np.datetime64(mask.last_train_timestamp) + np.timedelta64(4, "D")
+
+
+@pytest.mark.parametrize(
+    "lag,expected", ((5, {"segment_0": 76.923077, "segment_1": 90.909091}), (6, {"segment_0": 100, "segment_1": 120}))
+)
+def test_backtest(simple_ts: TSDataset, lag: int, expected: Dict[str, List[float]]):
+    mask = FoldMask(
+        simple_ts.index.min(),
+        simple_ts.index.min() + np.timedelta64(6, "D"),
+        [simple_ts.index.min() + np.timedelta64(8, "D")],
+    )
+    pipeline = Pipeline(model=NaiveModel(lag=lag), transforms=[], horizon=2)
+    metrics_df, _, _ = pipeline.backtest(ts=simple_ts, metrics=[SMAPE()], n_folds=[mask], aggregate_metrics=True)
+    metrics = dict(metrics_df.values)
+    for segment in expected.keys():
+        assert segment in metrics.keys()
+        np.testing.assert_array_almost_equal(expected[segment], metrics[segment])
+
+
+@pytest.mark.parametrize(
+    "lag,expected", ((4, {"segment_0": 0, "segment_1": 0}), (7, {"segment_0": 0, "segment_1": 0.5}))
+)
+def test_backtest_right_timestamps(masked_ts: TSDataset, lag: int, expected: Dict[str, List[float]]):
+    mask = FoldMask(
+        masked_ts.index.min(),
+        masked_ts.index.min() + np.timedelta64(6, "D"),
+        [masked_ts.index.min() + np.timedelta64(9, "D"), masked_ts.index.min() + np.timedelta64(10, "D")],
+    )
+    pipeline = Pipeline(model=NaiveModel(lag=lag), transforms=[], horizon=4)
+    metrics_df, _, _ = pipeline.backtest(ts=masked_ts, metrics=[MAE()], n_folds=[mask], aggregate_metrics=True)
+    metrics = dict(metrics_df.values)
+    for segment in expected.keys():
+        assert segment in metrics.keys()
+        np.testing.assert_array_almost_equal(expected[segment], metrics[segment])
+
+
+@pytest.mark.parametrize(
+    "mask,expected",
+    (
+        (FoldMask("2020-01-01", "2020-01-07", ["2020-01-10"]), {"segment_0": 0, "segment_1": 11}),
+        (FoldMask("2020-01-01", "2020-01-07", ["2020-01-08", "2020-01-11"]), {"segment_0": 95.5, "segment_1": 5}),
+    ),
+)
+def test_run_fold(ts_run_fold: TSDataset, mask: FoldMask, expected: Dict[str, List[float]]):
```

Comment:
Shouldn't it be before testing backtest method?

Reply:
Fixed, got the order from etna-utils)